src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -933,7 +933,53 @@ public static bool IsPointerType(this TypeSymbol type)
 
         internal static int FixedBufferElementSizeInBytes(this TypeSymbol type)
         {
-            return type.SpecialType.FixedBufferElementSizeInBytes();
+            var specialType = type.SpecialType;
+
+            if (specialType != SpecialType.None)
+            {
+                return specialType.FixedBufferElementSizeInBytes();
+            }
+
+            if (type.TypeKind != TypeKind.Struct)
+            {
+                // TODO: Add Diagnostic
+                return 0;
+            }
+
+            // TODO: Do we need to handle explicit sizing of structs (those that are
+            // marked with `StructLayout(LayoutKind.Sequential, Pack = <blah>, Size = <blah>)`
+
+            var members = type.GetMembers();
+
+            int sizeInBytes = 0;
+
+            foreach (var member in members)
+            {
+                if (member.Kind != SymbolKind.Field)
+                {
+                    continue;
+                }
+
+                var fieldSymbol = member as FieldSymbol;
+                var typeSymbol = fieldSymbol.Type;
+
+                if (fieldSymbol.IsFixed)
+                {
+                    typeSymbol = ((PointerTypeSymbol)typeSymbol).PointedAtType;
+                }
+
+                int elementSize = typeSymbol.FixedBufferElementSizeInBytes();
+
+                if (elementSize == 0)
+                {
+                    // TODO: Add Diagnostic
+                    return 0;
+                }
+
+                sizeInBytes += elementSize;
```

Comment:
Do we need to account for padding? For instance the size of `struct S { char c; int i; }` is probably 8.